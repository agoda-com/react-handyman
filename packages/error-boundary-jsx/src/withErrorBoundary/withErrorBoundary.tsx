/* eslint-disable react/jsx-props-no-spreading */
/* eslint-disable @typescript-eslint/no-explicit-any */
import * as React from 'react';
import ErrorBoundary from '../ErrorBoundary';

const withErrorBoundary = <TOrigProps extends Record<string, unknown>>(
  Component: React.ComponentType<TOrigProps>,
  onError: (componentName: string, error: Error, componentStack: string) => void,
  FallbackComponent?: React.ComponentType<any>
) => {
  const name = Component.displayName ? Component.displayName : 'Unnamed Component';

  const ErrorBoundaryWrapper = (props: TOrigProps) => (
    <ErrorBoundary onError={onError} name={name} FallbackComponent={FallbackComponent}>
      <Component {...props} />
    </ErrorBoundary>
  );

  ErrorBoundaryWrapper.displayName = `WithErrorBoundary(${name})`;

  return ErrorBoundaryWrapper;
};

export default withErrorBoundary;
