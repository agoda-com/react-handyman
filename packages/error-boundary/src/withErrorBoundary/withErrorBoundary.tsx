import * as React from 'react';

interface ErrorInfo {
  componentStack: string;
}

const withErrorBoundary = <TProps extends {}>(
  Component: React.ComponentType<TProps>,
  name: string,
  onError: (error: Error, componentName: string, componentStack: string) => void,
) => {

  return class ErrorBoundary extends React.PureComponent<TProps, {}> {

    static displayName = `withErrorBoundary(${Component.displayName})`;

    constructor(props: TProps) {
      super(props);
    }

    componentDidCatch(error: Error, info: ErrorInfo) {
      if (typeof onError === 'function') {
        try {
          onError.call(this, error, name, info ? info.componentStack : '');
        } catch (e) { }
      }
    }

    render(): JSX.Element {
      return <Component {...this.props} />;
    }
  };
}

export default withErrorBoundary;
