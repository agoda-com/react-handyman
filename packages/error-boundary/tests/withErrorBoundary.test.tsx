import 'jest';
import * as React from 'react';
import withErrorBoundary from '../src/withErrorBoundary';
import { render } from '@testing-library/react';

interface DummyErrorComponentProps {
  name?: string;
}
const defaultProps: DummyErrorComponentProps = {
  name: 'Test name',
};

class DummyErrorComponent extends React.PureComponent<DummyErrorComponentProps, {}> {
  constructor(props: DummyErrorComponentProps) {
    super(props);
    throw Error('error');
  }
  render() {
    return <div>Dummy Component with constructor error</div>;
  }
}

describe('<ErrorBoundary />', () => {

  const errorCallback = jest.fn();

  it('should call componentDidCatch on error and fire callback', () => {

    const ComponentWithError = withErrorBoundary(DummyErrorComponent, 'component name', errorCallback);

    jest.spyOn(ComponentWithError.prototype, 'componentDidCatch');
    render(<ComponentWithError {...defaultProps} />);
    expect(ComponentWithError.prototype.componentDidCatch).toHaveBeenCalledTimes(1);
    expect(errorCallback).toHaveBeenCalledTimes(1);
  });
});
