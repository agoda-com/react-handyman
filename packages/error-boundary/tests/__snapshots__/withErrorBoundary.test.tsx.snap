// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<withErrorBoundary /> should render custom fallback component on error 1`] = `
<DocumentFragment>
  <div>
    <h3>
      error
    </h3>
    <span>
      
    in BuggyComponent (created by WithErrorBoundary(BuggyComponent))
    in ErrorBoundary (created by WithErrorBoundary(BuggyComponent))
    in WithErrorBoundary(BuggyComponent)
    </span>
  </div>
</DocumentFragment>
`;

exports[`<withErrorBoundary /> should render default fallback component on error 1`] = `
<DocumentFragment>
  <h2>
    Something went wrong.
  </h2>
</DocumentFragment>
`;
